runtime/codert_vm/cnathelp.cpp

```diff
@@ -3240,12 +3240,16 @@ old_slow_jitNewInstanceImplAccessCheck(J9VMThread *currentThread)
 			 * unless the two classes are in the same nest (JDK11 and beyond).
 			 */
 #if JAVA_SPEC_VERSION >= 11
-			if (NULL == thisClass->nestHost) {
+			if ((NULL == thisClass->nestHost)
```

Comment:
The real problem is that `checkVisibility` can call Java code and therefore release VMAccess allowing HCR to occur. 

The normal way we address this is to do:

```
VM_VMHelpers::currentClass(thisClass);
VM_VMHelpers::currentClass(callerClass);
```

after the checkVisibility call

Reply:
OK, I guess we need to call `VM_VMHelpers::currentClass()` to update `thisClass`, `callerClass` as well as `thisClass->nextHost`, `callerClass->nextHost` ?

Reply:
I added `thisClass = VM_VMHelpers::currentClass(thisClass)` and `callerClass = VM_VMHelpers::currentClass(callerClass)` after `checkVisibility()`. The illegal access error is still there.

Maybe I should point out this piece of code that updates the nesthost of all nest members in class redefinition:

https://github.com/eclipse-openj9/openj9/blob/a2c8e150ded25bc8bd118e4b9b76653ec5aa40d8/runtime/util/hshelp.c#L2421-L2429

The above code does not work for nest member that is a hidden class. Hidden class is added to the nest at runtime. It is not in the nestmember attribute of the nesthost. The nesthost cannot reference its hidden class nest member by name either. In this issue, `thisClass` is a hidden class, it is never redefined. `thisClass->nextHost` still points to the obsolete `callerClass` after `callerClass` is redefined (and the calls of `VM_VMHelpers::currentClass()` on `thisClass` and `callerClass`).